### PR TITLE
More glibc allocation removals.

### PR DESCRIPTION
**Pull request by Sonicadvance1**
~~First 39 commits are from #2549 and that needs to be merged first.~~

This hits a whole bunch more things that allocate memory through the glibc allocators.

~~#2583 must be merged first otherwise this will break thunks.~~

### CODE REVIEW

File: FEXHeaderUtils/FEXHeaderUtils/Filesystem.h [hunk 1]
```diff
@@ -200,6 +200,18 @@ namespace FHU::Filesystem {
     return false;
   }
 
+  /**
+   * @brief Renames a file and overwrites if it already exists.
+   *
+   * @param From
+   * @param To
+   *
```
Comment (neobrain) [inline FEXHeaderUtils/FEXHeaderUtils/Filesystem.h:208]: These aren't useful, so they can be removed.

File: Source/Tests/AOT/AOTGenerator.cpp [hunk 2]
```diff
@@ -144,6 +147,7 @@ void AOTGenSection(FEXCore::Context::Context *CTX, ELFCodeLoader::LoadedSection
 
       // All entryproints processed, cleanup this thread
       CTX->DestroyThread(Thread);
+      FEXCore::Allocator::YesIKnowImNotSupposedToUseTheGlibcAllocator::HardDisable();
```
Comment (neobrain) [inline Source/Tests/AOT/AOTGenerator.cpp:150]: Should have a comment explaining why it's used.

File: ThunkLibs/libxcb/libxcb_Guest.cpp [hunk 3]
```diff
@@ -30,6 +30,8 @@ tags: thunklibs|xcb
 
 static std::thread CBThread{};
 static std::atomic<bool> CBDone{false};
+static std::mutex CBThreadMutex;
```
Comment (neobrain) [inline ThunkLibs/libxcb/libxcb_Guest.cpp:33]: Are these changes related to the other changes in this PR? If not, would be better suited in a dedicated PR since it's already a fairly large bunch of changes as-is.

File: External/FEXCore/Scripts/config_generator.py [hunk 4]
```diff
@@ -371,13 +371,16 @@ def print_parse_argloader_options(options):
...
                 NeedsString = True
-                conversion_func = "FEXCore::Config::Handler::{0}".format(op_vals["ArgumentHandler"])
+                conversion_func = "FEXCore::Config::Handler::{0}(".format(op_vals["ArgumentHandler"])
             if (value_type == "str"):
                 NeedsString = True
-                conversion_func = ""
+                conversion_func = "("
+            if (value_type == "bool"):
+                # boolean values need a decimal specifier. Otherwise fmt prints strings.
+                conversion_func = "fextl::fmt::format(\"{:d}\", "
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Scripts/config_generator.py:383]:
```cpp
if (Options.is_set_by_user("LibraryJITNaming")) {
  bool UserValue = Options.get("LibraryJITNaming");
  Set(FEXCore::Config::ConfigOption::CONFIG_LIBRARYJITNAMING, +UserValue));
}
```
compile output
```
/mnt/Work/Work/work/FEXNew/Build/include/FEXCore/Config/ConfigOptions.inl:591:72: error: extraneous ')' before ';'
        Set(FEXCore::Config::ConfigOption::CONFIG_GLOBALJITNAMING, +UserValue));
                                                                              ^
/mnt/Work/Work/work/FEXNew/Build/include/FEXCore/Config/ConfigOptions.inl:595:62: error: no viable conversion from 'int' to 'std::string_view' (aka 'basic_string_view<char>')
        Set(FEXCore::Config::ConfigOption::CONFIG_LIBRARYJITNAMING, +UserValue));
```

???

File: External/FEXCore/Source/Interface/Core/GdbServer.cpp [hunk 5]
```diff
@@ -1309,17 +1308,19 @@ void GdbServer::OpenListenSocket() {
   }
 
   listen(ListenSocket, 1);
+
+  freeaddrinfo(res);
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/GdbServer.cpp:1312]: Was just leaking memory prior to this.

File: External/FEXCore/Source/Interface/IR/AOTIR.cpp [hunk 6]
```diff
@@ -200,6 +198,9 @@ namespace FEXCore::IR {
     }
 
     for (;;) {
+      // This code is tricky to refactor so it doesn't allocate memory through glibc.
+      FEXCore::Allocator::YesIKnowImNotSupposedToUseTheGlibcAllocator glibc;
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/IR/AOTIR.cpp:202]: The entire std::function object allocates memory and deallocates once the object reaches end of scope.

Adding comment.

File: FEXHeaderUtils/FEXHeaderUtils/Filesystem.h [hunk 7]
```diff
@@ -200,6 +200,18 @@ namespace FHU::Filesystem {
...
+   * @return True if the rename occured, False otherwise.
+   */
+  inline bool RenameFile(const fextl::string &From, const fextl::string &To) {
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply FEXHeaderUtils/FEXHeaderUtils/Filesystem.h:211]: As in void return type?

File: FEXHeaderUtils/FEXHeaderUtils/Filesystem.h (hunk 1 above)
Comment (Sonicadvance1) [reply FEXHeaderUtils/FEXHeaderUtils/Filesystem.h]: Removed.

File: External/FEXCore/include/FEXCore/Core/Context.h [hunk 8]
```diff
@@ -70,6 +70,16 @@ namespace FEXCore::Context {
       std::unique_lock<std::shared_mutex> lock;
   };
 
+  class AOTIRWriterFD {
+    public:
+      virtual ~AOTIRWriterFD() {
+      }
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/include/FEXCore/Core/Context.h]: Changed.

File: External/FEXCore/Source/Interface/IR/AOTIR.cpp [hunk 9]
```diff
@@ -327,6 +328,9 @@ namespace FEXCore::IR {
           auto RADataCopy = RAData->CreateCopy();
           auto RADataCopyDeleter = RADataCopy.get_deleter();
           auto IRListCopy = IRList->CreateCopy();
+
+          // This code is tricky to refactor so it doesn't allocate memory through glibc.
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/IR/AOTIR.cpp]: Changed.

File: External/FEXCore/Source/Interface/Core/GdbServer.cpp [hunk 10]
```diff
@@ -1280,7 +1279,7 @@ void GdbServer::StartThread() {
 }
 
 void GdbServer::OpenListenSocket() {
-  // open socket
+  FEXCore::Allocator::YesIKnowImNotSupposedToUseTheGlibcAllocator glibc;
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/GdbServer.cpp:1283]: Added a comment.

File: Source/Tests/FEXLoader.cpp [hunk 11]
```diff
@@ -455,17 +452,48 @@ int main(int argc, char **argv, char **const envp) {
...
-      auto filepath = std::filesystem::path(FEXCore::Config::GetDataDirectory()) / "aotir" / (fileid + ".aotir");
-
+      const auto filepath = fextl::fmt::format("{}/aotir/{}.aotir", FEXCore::Config::GetDataDirectory(), fileid);
       return open(filepath.c_str(), O_RDONLY);
     });
 
-    CTX->SetAOTIRWriter([](const fextl::string& fileid) -> std::unique_ptr<std::ofstream> {
-      auto filepath = std::filesystem::path(FEXCore::Config::GetDataDirectory()) / "aotir" / (fileid + ".aotir.tmp");
-      auto AOTWrite = std::make_unique<std::ofstream>(filepath, std::ios::out | std::ios::binary);
+    class AOTIRWriterFD final : public FEXCore::Context::AOTIRWriterFD {
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply Source/Tests/FEXLoader.cpp]: Moved elsewhere. At some point this AOT IR stuff is just going to get completely shuffled around since it's a mess.

File: ThunkLibs/libxcb/libxcb_Guest.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply ThunkLibs/libxcb/libxcb_Guest.cpp:33]: It's related to this PR in that this PR completely breaks xcb thunking if this isn't in place.
We could move it out to a PR that needs to be merged before this one?

File: External/FEXCore/Scripts/config_generator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neobrain) [reply External/FEXCore/Scripts/config_generator.py:383]: Ah, this is just missing the opening parenthesis, but also: I thought the output of `conversion_func` was itself fed into another call to `fmt::format`, so I wanted to fix a double-format.

We could change the line to `conversion_func += '+';` instead, but not sure it's that much better as far as readability goes.

[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply External/FEXCore/Scripts/config_generator.py:383]: Nope, Set just consumes a string_view and makes a copy internally. Which with some larger refactoring could take all the different types and convert internally but that's out of scope.

Seems easier to read with the original route?

File: ThunkLibs/libxcb/libxcb_Guest.cpp (hunk 3 above)
Comment (neobrain) [reply ThunkLibs/libxcb/libxcb_Guest.cpp:33]: Different PR sounds good, just to make it clear this one really just handles `std->fextl` migration itself. Nothing egregious about the change itself so it should be an easy merge.

File: External/FEXCore/Scripts/config_generator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neobrain) [reply External/FEXCore/Scripts/config_generator.py:383]: Yeah, I don't mind either way now that I understand the logic right. Current code is fine.

File: ThunkLibs/libxcb/libxcb_Guest.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply ThunkLibs/libxcb/libxcb_Guest.cpp]: #2583 opened and that commit dropped from this PR. That PR needs to be merged first.

File: External/FEXCore/Source/Interface/IR/AOTIR.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (neobrain) [reply External/FEXCore/Source/Interface/IR/AOTIR.cpp:203]: Are you sure the function itself allocates memory here? It's using move construction, so the allocation should already have happened when appending to AOTIRCaptureCacheWriteoutQueue.

For deallocation you're right though. And it means we can't move the scope-guard down any further either since it needs to live longer than `fn` :(

So the comment needs a minor update for this unless I'm missing something else about the allocation: `The moved std::function object deallocates memory at the end of scope.` would be fine.

File: FEXHeaderUtils/FEXHeaderUtils/Filesystem.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (neobrain) [reply FEXHeaderUtils/FEXHeaderUtils/Filesystem.h:208]: Yes, void return type + `std::error_code` parameter. Returning a `[[nodiscard]] std::error_code` would be fine too, but both are explicit enough when discarding error codes.

[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply FEXHeaderUtils/FEXHeaderUtils/Filesystem.h]: I let it return a std::error_code.

File: Source/Tests/FEXLoader.cpp [hunk 12]
```diff
@@ -299,7 +333,7 @@ int main(int argc, char **argv, char **const envp) {
   if (!ExecutedWithFD && !FEXFD && !FHU::Filesystem::Exists(Program.ProgramPath)) {
     // Early exit if the program passed in doesn't exist
     // Will prevent a crash later
-    fmt::print(stderr, "{}: command not found\n", Program.ProgramPath);
+    fextl::fmt::print("{}: command not found\n", Program.ProgramPath);
```
Comment (neobrain) [inline Source/Tests/FEXLoader.cpp:336]: Why is `stderr` dropped here and in the other uses in this file? Don't we just need another `print` overload to support the original semantics?

**Review by neobrain — APPROVED** [review]

File: Source/Tests/FEXLoader.cpp (hunk 12 above)
Comment (Sonicadvance1) [reply Source/Tests/FEXLoader.cpp]: Resolved by adding FILE handlers just like fmt.
Will need to check later if fwrite itself does memory allocations but for the sake of pushing this through I'll let it be for now.
